### PR TITLE
Bug 1498618 - Support bind parameters.

### PR DESCRIPTION
**Pull request by cfchase**
Support bind parameters. Proposal included.

### CODE REVIEW

**Comment by cfchase** [comment]
[(8) Bind Parameters](https://trello.com/c/DXrXuPnQ/525-8-bind-parameters)

File: docs/proposals/bind-parameters.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+## ASB Changes
+ASB will parse the `bind_parameters` and add them to the plan.  The catalog request should use the `bind_parameters` and send them as part of the service instances, allowing the OpenShift UI to use them.  The form definitions must also be supplied for the UI.  When a binding is created, ASB should capture the parameters from the request and send them to the bind APB as `extra-vars` available for use.
+
+## Work Items
+* ASB Changes
+* ASB Unit tests
+* OpenShift UI form definition changes
+* Experimental APB with async bind and passed parameters
+* APB documentation changes (getting started / developer guide)
```
Comment (jmrodri) [inline docs/proposals/bind-parameters.md]: Normally the proposal is a separate PR done before the implementation :) But the proposal seems sane and it is very clear of your intentions. +1

File: pkg/apb/types.go [hunk 2]
```diff
@@ -50,12 +50,13 @@ type ParameterDescriptor struct {
...
-	Free        bool                   `json:"free,omitempty"`
-	Bindable    bool                   `json:"bindable,omitempty"`
-	Parameters  []ParameterDescriptor  `json:"parameters"`
+	Name           string                 `json:"name"`
+	Description    string                 `json:"description"`
+	Metadata       map[string]interface{} `json:"metadata,omitempty"`
+	Free           bool                   `json:"free,omitempty"`
+	Bindable       bool                   `json:"bindable,omitempty"`
+	Parameters     []ParameterDescriptor  `json:"parameters"`
+	BindParameters []ParameterDescriptor  `json:"bind_parameters,omitempty" yaml:"bind_parameters,omitempty"`
```
Comment (jmrodri) [inline pkg/apb/types.go]: 👍

File: pkg/broker/broker.go [hunk 3]
```diff
@@ -677,11 +677,11 @@ func (a AnsibleBroker) Bind(instance apb.ServiceInstance, bindingUUID uuid.UUID,
 	// See if the service instance still exists, if not send back a badrequest.
 
 	// GET SERVICE get provision parameters
-	params := make(apb.Parameters)
-	if instance.Parameters != nil {
-		params["provision_params"] = *instance.Parameters
+	params := req.Parameters
+	if params == nil {
+		params = make(apb.Parameters)
 	}
```
Comment (jmrodri) [inline pkg/broker/broker.go]: 👍

File: pkg/broker/util.go [hunk 4]
```diff
@@ -171,10 +200,44 @@ func getType(paramType string) schema.PrimitiveTypes {
 	return []schema.PrimitiveType{schema.UnspecifiedType}
 }
 
-func parametersToSchema(params []apb.ParameterDescriptor) Schema {
+func parametersToSchema(plan apb.Plan) Schema {
```
Comment (jmrodri) [inline pkg/broker/util.go]: Sane change to parametersToSchema. 👍

File: pkg/broker/util_test.go [hunk 5]
```diff
@@ -76,18 +76,43 @@ var PlanParams = []apb.ParameterDescriptor{
 	},
 }
 
+var PlanBindParams = []apb.ParameterDescriptor{
+	{
```
Comment (jmrodri) [inline pkg/broker/util_test.go]: YES! to updating tests. 👏

**Comment by djzager** [comment]
Added the `do-not-merge` to hold onto this until after we have branched for 3.7 release.

**Pull request merged**